### PR TITLE
[Dev] Adding prometheus and grafana as a service

### PR DESCRIPTION
**Pull request by gabrielifg**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

feature

* **What is the current behavior?** (You can also link to an open issue here)

There is currently no support for prometheus and grafana on the DojoT platform with docker compose.

* **What is the new behavior (if this is a feature change)?**

Now we have support for prometheus and grafana on the DojoT platform with docker compose.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

No

* **Is there any issue related to this PR in other repository?** (such as dojot/dojot)

dojot/dojot#2201

* **Other information**:

### CODE REVIEW

File: monitoring/grafana/provisioning/dashboards/dashboards.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+
+apiVersion: 1
+
+providers:
+- name: 'Prometheus'
+  orgId: 1
+  folder: 'DojoT'
```
Comment (marianoleonardo) [inline monitoring/grafana/provisioning/dashboards/dashboards.yml:7]: I think you could change `DojoT` to `dojot`

File: monitoring/grafana/provisioning/datasources/datasource.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# # config file version
```
Comment (marianoleonardo) [inline monitoring/grafana/provisioning/datasources/datasource.yml:1]: This file is OK, but i think you should remove the comments to make it cleanner.

File: monitoring/prometheus/prometheus.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  - job_name: cadvisor
+    scrape_interval: 5s
+    static_configs:
+      - targets: ['cadvisor:8080']
+  - job_name: node_exporter
+    static_configs:
+      - targets: ['node_exporter:9100']
+  - job_name: docker
+    static_configs:
+      - targets: ['172.17.0.1:9323'] # 172.17.0.1 porta padrão da interface de rede do docker "docker0"
```
Comment (marianoleonardo) [inline monitoring/prometheus/prometheus.yml:33]: I think you should translate this part to english.

File: monitoring/prometheus/prometheus.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+scrape_configs:
+  - job_name: prometheus
+    metrics_path: /metrics
+    honor_labels: false
+    honor_timestamps: true
+    scheme: http
+    #sample_limit: 0
+    static_configs:
+      - targets: ['localhost:9090'] #, '192.168.240.27:8888']
+  - job_name: cadvisor
```
Comment (marianoleonardo) [inline monitoring/prometheus/prometheus.yml:24]: I think you could add a "enter" here to make it easier to distinguish the jobs

File: monitoring/prometheus/prometheus.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    scrape_interval: 5s
+    static_configs:
+      - targets: ['cadvisor:8080']
+  - job_name: node_exporter
```
Comment (marianoleonardo) [inline monitoring/prometheus/prometheus.yml:28]: I think you could add a "enter" here to make it easier to distinguish the jobs

File: monitoring/prometheus/prometheus.yml [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    static_configs:
+      - targets: ['node_exporter:9100']
+  - job_name: docker
```
Comment (marianoleonardo) [inline monitoring/prometheus/prometheus.yml:31]: I think you could add a "enter" here to make it easier to distinguish the jobs

File: monitoring/prometheus/prometheus.yml [hunk 7]
```diff
@@ -0,0 +1,43 @@
+global:
```
Comment (marianoleonardo) [inline monitoring/prometheus/prometheus.yml:1]: I think you need to remove some comments as you did in datasource.yaml to make the file cleanner.

**Review by marianoleonardo — CHANGES_REQUESTED** [review]

File: monitoring/PROMETHEUS.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+* Used space of available disks.
+
+
+## CAdvisor
+
+CAdvisor is an export developed by google providing metrics regarding resource usage and performance characteristics of running containers.
+
+Following the same line of reasoning presented for Node-Exporter, below are the metrics we are analyzing:
+
+* Current status of the container: Running, Stopped and Paused;
```
[the remark this answers is not in the document]
Comment (gabrielifg) [reply monitoring/PROMETHEUS.md:37]: Nesse caso Leo, de fato é "and" mesmo, por conta da estrutura no grafana. Lá nos analisamos essas informações em panels separados e não de forma unitária onde você pode ver um "ou" outro.

File: monitoring/PROMETHEUS.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+* Network traffic received by container;
+* Network traffic sent by container;
+* Percentage of memory usage per container;
+* Percentage of CPU usage per container;
+* RSS Memory Usage per Container (Among the memory allocated to the process, how much is being used by RAM memory)
+* Use of memory RAM per container;
```
[the remark this answers is not in the document]
Comment (gabrielifg) [reply monitoring/PROMETHEUS.md:43]: Sim sim, a métrica é semelhante mas no dashboard elas são apresentadas de formas diferentes. No primeiro caso, em gráficos e no segundo, em tabelas. No fim, a abordagem é diferente.

File: monitoring/PROMETHEUS.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+We are currently extracting the following metrics with CAdvisor:
+
+
+* Current status of the container: ``Running``, ``Stopped`` and ``Paused``;
+* Network traffic received by container;
+* Network traffic sent by container;
+* Percentage of RAM usage per container;
+* Percentage of CPU usage per container;
+* RSS Memory Usage per Container
```
Comment (marianoleonardo) [inline monitoring/PROMETHEUS.md:43]: You should write a `;` at the end of the text

File: monitoring/PROMETHEUS.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+* RAM consumption table per container;
+* Available RAM for each conatainer;
+*  RAM limit established for each container
```
Comment (marianoleonardo) [inline monitoring/PROMETHEUS.md:46]: You should write a `.` at the end of the text

File: monitoring/PROMETHEUS.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+If the file is empty, paste the following:
+
+``
+{
+  "metrics-addr" : "0.0.0.0:9323",
+  "experimental" : true
+}
+``
+
+If the file is not empty, add these two keys, making sure the resulting file is a valid JSON. Be careful that all lines end with a comma  "," except the last line.
```
Comment (marianoleonardo) [inline monitoring/PROMETHEUS.md:65]: There is two spaces between `comma` and `","`. You should remove one.
i think you should add a `must` between `lines` and `end` like this:
```
Be careful that all lines must end...
```

File: monitoring/PROMETHEUS.md [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+```
+cd monitoring
+
+docker-compose -f docker-compose-monitoring.yml up -d
+```
+
+When the process is complete, we can check whether the services have been started:
+
+*   [Metrics Docker](http://localhost:9323/metrics).
```
Comment (marianoleonardo) [inline monitoring/PROMETHEUS.md:113]: Change to `Docker Metrics`

File: monitoring/PROMETHEUS.md [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+*   [Metrics Prometheus](http://localhost:9090/metrics).
```
Comment (marianoleonardo) [inline monitoring/PROMETHEUS.md:115]: Change to `Prometheus Metrics`

File: monitoring/PROMETHEUS.md [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+Visually, it is possible to check if everything is "UP" by [Targets Prometheus](http://localhost:9090/targets).
```
Comment (marianoleonardo) [inline monitoring/PROMETHEUS.md:117]: Change to `Prometheus Targets`

File: monitoring/grafana/provisioning/dashboards/dashboards.yml [hunk 16]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (marianoleonardo) [inline monitoring/grafana/provisioning/dashboards/dashboards.yml:1]: I think you could remove this extra line in the beginning of the file.

**Review by marianoleonardo — CHANGES_REQUESTED** [review]
I left some comments

File: monitoring/PROMETHEUS.md [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+``
+{
+  "metrics-addr" : "0.0.0.0:9323",
+  "experimental" : true
+}
+``
+
+If the file is not empty, add these two keys, making sure the resulting file is a valid JSON. Be careful that all lines end with a comma ( ,) except the last line.
+
+Since we are running prometheus on localhost, in case we try to use the docker as localhost as well. Prometheus would understand that the metrics would be inside the service itself (prometheus) and then try to access port ``9323`` inside the promethues container. To prevent this from happening we will use the default port **172.17.0.1** of the **bridge** docker ``docker0``.
```
[the remark this answers is not in the document]
Comment (marianoleonardo) [reply monitoring/PROMETHEUS.md]: I think you need to make this changes.

**Review by marianoleonardo — APPROVED** [review]
Great job! 👏🏼👏🏼

**Review by danielapetito — APPROVED** [review]

**Pull request merged**